dashboard/src/actions/authActions.js

```diff
@@ -7,6 +7,161 @@ import Cookies from "js-cookie";
+          const response = JSON.parse(reqUser.responseText);
+          userPayload = {
+            "username": response?.preferred_username,
+            "email": response?.email,
+            "first_name": response?.given_name,
+            "last_name": response?.family_name,
+          };
+          resolve(userPayload);
+        }
+      }
+      reqUser.open('GET', uri, false);
+      reqUser.setRequestHeader('Content-type', 'application/json');
+      reqUser.setRequestHeader('Authorization', 'Bearer ' + OIDCToken);
+      reqUser.send();
+     });
+}
+
+export const loadTokens = () => async (dispatch, getState) => {
+    const code = getQueryVariable('code');
+
+    if (code !== "") {
+      let endpoints = getState().apiEndpoint.endpoints;
+      if (!("openid" in endpoints)){
+          await dispatch(fetchEndpoints);
+      }
+      endpoints = getState().apiEndpoint.endpoints;
+
+      const oidcServer = endpoints["openid"]?.server
+      const oidcRealm = endpoints["openid"]?.realm
+      const oidcClient = endpoints["openid"]?.client
+      const uri = `${oidcServer}/realms/${oidcRealm}/protocol/openid-connect/token`;
+      const queryParams = [
+        'grant_type=authorization_code',
+        'client_id=' + oidcClient,
+        'code=' + code,
+        'redirect_uri=' + window.location.href.split('?')[0],
+      ];
+
+      const req = new XMLHttpRequest();
+      req.onreadystatechange = async function () {
+        if (req.readyState === 4) {
+          const response = JSON.parse(req.responseText);
+          const keepUser = getState().userAuth.keepLoggedIn;
+          const expiryTime = keepUser
+            ? CONSTANTS.EXPIRY_KEEPUSER_DAYS
+            : CONSTANTS.EXPIRY_DEFAULT_DAYS;
+          Cookies.set("isLoggedIn", true, { expires: expiryTime });
+          Cookies.set("token", response["access_token"], {
+            expires: response["expires_in"],
+          });
+          let userPayload = {};
+          await getOIDCUserDetails(endpoints).then(res => userPayload = res);
+          dispatch({
+            type: TYPES.GET_USER_DETAILS,
+            payload: userPayload,
+          });
+          Cookies.set("refresh_token", response["refresh_token"], {
+            expires: response["refresh_expires_in"],
+          });
+          Cookies.set("username", userPayload.username, {
+            expires: expiryTime,
+          });
+          const loginDetails = {
+            isLoggedIn: true,
+            token: response["access_token"],
+            username: userPayload.username,
+          };
+          dispatch({
+            type: TYPES.SET_LOGIN_DETAILS,
+            payload: loginDetails,
+          });
+          dispatch(showToast(SUCCESS, "Logged in successfully!"));
+        }
+        dispatch({ type: TYPES.COMPLETED });
+      }
+      req.open('POST', uri);
+      req.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
+      req.send(queryParams.join('&'));
```

Comment:
You're supposed to be able to send a [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) object rather than manually stitching together a string. Also, do we really want to use the outdated `XMLHttpRequest` object rather than the rather more modern `fetch` or the axios package, which we're using elsewhere for HTTP requests?